### PR TITLE
Rewrite tests using vscode-jsonrpc

### PR DESCRIPTION
**Pull request by remcohaszing**
### Initial checklist

*   [x] I read the support docs 
*   [x] I read the contributing guide 
*   [x] I agree to follow the code of conduct 
*   [x] I searched issues and couldn’t find anything (or linked relevant results below) 
*   [x] If applicable, I’ve added docs and tests

### Description of changes

This way of testing doesn’t require to use any delays or timeouts.

This also works by sending a response, awaiting a reply, then asserting the reply, instead of sending a number of requests only to assert all responses together later.

Also any logging messages sent from the connection.console are logged to the console in tests.

It’s also less boilerplate per test.

All of these changes together make troubleshooting any tests much nicer to deal with.

This is a draft because I’d like to have some feedback on this testing method and rebase this on #33 before reworking all tests.

### CODE REVIEW

**Review by wooorm — COMMENTED** [review]
Good idea!

Sometimes you do:

```
const xxxResponse = await xxx()

t.deepEqual(
  xxxResponse,
```

Other times its:

```
t.deepEqual(
  await xxx(),
```

I think its good to be consistent there.
While the `await` inside an assertion means one less variable, I do find it harder to track what’s sync and what’s async and how the program progresses.

---

The flow as noted above might also be helped by improving the names of variables:

```
const openDiagnostics = createDiagnosticsPromise(connection)

// …

t.deepEqual(
  await openDiagnostics,
```

From the name, it sounds like `openDiagnostics` is actually a value, not a promise.
Calling it `promise` or `diagnosticsPromise` (or if the `open` is needed here `openDiagnosticsPromise`)

Both together:

```
const diagnosticsPromise = createDiagnosticsPromise(connection)

// …

const diagnostics = await diagnosticsPromise

t.deepEqual(
  diagnostics,
```

---

From my work on this project, I experienced that sometimes notifications don’t work. Should `createDiagnosticsPromise` therefore have a timeout, that rejects?

---

Should `createDiagnosticsPromise`, when the rest is converted as well, be made more generic, in that it can handle different events?

---

In this code:

```
const openDiagnostics = createDiagnosticsPromise(connection)
connection.sendNotification(
  'textDocument/didOpen',
// …
t.deepEqual(
  await openDiagnostics,
```

…isn’t it weird that the first line sends `'textDocument/publishDiagnostics'`, and the second line immediately sends `'textDocument/didOpen'`, and then *the first* is awaited?
Shouldn’t the `'textDocument/didOpen'` be sent first (and awaited, to prevent race conditions), and *then* `'textDocument/publishDiagnostics'` be send last (and awaited?)

**Comment by remcohaszing** [comment]
> The flow as noted above might also be helped by improving the names of variables:
> 
> ```
> const openDiagnostics = createDiagnosticsPromise(connection)
> 
> // …
> 
> t.deepEqual(
>   await openDiagnostics,
> ```
> 
> From the name, it sounds like `openDiagnostics` is actually a value, not a promise. Calling it `promise` or `diagnosticsPromise` (or if the `open` is needed here `openDiagnosticsPromise`)
> 
> Both together:
> 
> ```
> const diagnosticsPromise = createDiagnosticsPromise(connection)
> 
> // …
> 
> const diagnostics = await diagnosticsPromise
> 
> t.deepEqual(
>   diagnostics,
> ```

I agree :+1: 

---

> From my work on this project, I experienced that sometimes notifications don’t work. Should `createDiagnosticsPromise` therefore have a timeout, that rejects?

Notifications work reliably (see explanation below), but still adding a timeout is useful. I’ll use `t.timeoutAfter(2000)` (maybe tweak the value as needed).

> ---
>
> Should `createDiagnosticsPromise`, when the rest is converted as well, be made more generic, in that it can handle different events?

Maybe if this turns out to be necessary. The way it’s written now is also type safe for this specific use case. I don’t want to overengineer the tests to accommodate scenarios that are currently unused.

> ---
>
> In this code:
> 
> ```
> const openDiagnostics = createDiagnosticsPromise(connection)
> connection.sendNotification(
>   'textDocument/didOpen',
> // …
> t.deepEqual(
>   await openDiagnostics,
> ```
> 
> …isn’t it weird that the first line sends `'textDocument/publishDiagnostics'`, and the second line immediately sends `'textDocument/didOpen'`, and then _the first_ is awaited? Shouldn’t the `'textDocument/didOpen'` be sent first (and awaited, to prevent race conditions), and _then_ `'textDocument/publishDiagnostics'` be send last (and awaited?)

No, but your explanation shows why you’ve encountered timeout issues before.

There are 3 types of communication between the client (tests) and server.

1. methods (`initialize`)
2. notifications from client to server (`textDocument/didOpen`)
3. notifications from server to client (`textDocument/publishDiagnostics`)

From the perspective of the tests:

- Methods can be awaited.
- Client notification listeners can be awaited, assuming they are fired.
- Sending notifications to the server is a synchronous action that should probably resolve a previously registered listener.

Registering the client notification listener after sending the notification from the client to the server causes the race condition.

**Comment by wooorm** [comment]
> No, but your explanation shows why you’ve encountered timeout issues before.

You are correct that I just messed up that one is a `sendNotification` and the other an `onNotification`. I don’t think that’s the reason why I experienced race conditions. Perhaps you’ll experience it too when rewriting the rest.

**Comment by wooorm** [comment]
I accidentally reviewed most of this in https://github.com/unifiedjs/unified-language-server/pull/37

File: test/index.js [hunk 1]
```diff
@@ -26,191 +36,90 @@ test('exports', (t) => {
...
+    ({
+      textDocument: {
+        uri,
+        languageId: 'markdown',
+        version: 1,
+        text: '# hi'
       }
     })
   )
+  const openDiagnostics = await openDiagnosticsPromise
```
[the remark this answers is not in the document]
Comment (wooorm) [reply test/index.js:115]: Yeah, I’d go with as few abstraction layers around the system we test! So, what do you prefer, current, or one of your two proposals?

**Comment by remcohaszing** [comment]
I’m closing this in favour of #37

**Pull request closed without merging**